pkg/exchange/binance/stream.go

```diff
@@ -146,36 +182,61 @@ func (s *Stream) handleDisconnect() {
 }
 
 func (s *Stream) handleConnect() {
+	// dboy: I think we don't need this anymore since we are removing listenKey(?)
+	// if !s.PublicOnly {
+	// 	// Emit Auth before establishing the connection to prevent the caller from missing the Update data after
+	// 	// creating the order.
+
+	// 	// spawn a goroutine to emit auth event to prevent blocking the main event loop
+	// 	go s.EmitAuth()
+	// 	return
+	// }
+	var cmd interface{}
 	if !s.PublicOnly {
-		// Emit Auth before establishing the connection to prevent the caller from missing the Update data after
-		// creating the order.
-
-		// spawn a goroutine to emit auth event to prevent blocking the main event loop
-		go s.EmitAuth()
-		return
-	}
-
-	var params []string
-	for _, subscription := range s.Subscriptions {
-		params = append(params, convertSubscription(subscription))
-	}
+		signature, timestamp := s.generateSignature()
+		cmd = LoginCommand{
+			WebSocketCommand: WebSocketCommand{
+				Method: "session.login",
+				ID:     1,
+			},
+			Params: LoginParams{
+				APIKey:    s.client.APIKey,
+				Signature: signature,
+				Timestamp: timestamp,
+			},
+		}
+	} else {
+		var params []string
+		for _, subscription := range s.Subscriptions {
+			params = append(params, convertSubscription(subscription))
+		}
 
-	if len(params) == 0 {
-		return
+		if len(params) == 0 {
+			return
+		}
+		cmd = WebSocketCommand{
+			Method: "SUBSCRIBE",
+			Params: params,
+			ID:     1,
+		}
+		log.Infof("subscribing channels: %+v", params)
 	}
-
-	log.Infof("subscribing channels: %+v", params)
-	err := s.Conn.WriteJSON(WebSocketCommand{
-		Method: "SUBSCRIBE",
-		Params: params,
-		ID:     1,
-	})
+	err := s.Conn.WriteJSON(cmd)
 
 	if err != nil {
 		log.WithError(err).Error("subscribe error")
 	}
 }
 
+func (s *Stream) generateSignature() (string, int64) {
```

Comment:
I guess you need to pull out this method as a pure function that takes one parameter -- the private key, because you will need to use it later for the RESTful API client.

Reply:
Yes.
I didn't think of backward compatibility.
I'll revise the PR as an whole.

Reply:
done.

Reply:
I move it to `pkg/exchange/coinbase/util.go` so the RESTful API can share this function with the stream.